### PR TITLE
Enable VPA functionality 

### PR DESCRIPTION
**Pull request by juliababkina**
Currently Skenario supports HPA functionality and is redesigned to be able to support VPA. 
This changes couples Skenario with VPA and turns vertical scaling in Skenario on.

### CODE REVIEW

File: plugin-k8s-vpa/pkg/plugin/plugin.go [hunk 1]
```diff
@@ -283,7 +283,7 @@ func (a *Autoscaler) VerticalRecommendation(now int64) ([]*proto.RecommendedPodR
 	//	})
 	//}
 	//TODO remove this code after fulfilling https://github.com/pivotal/skenario/issues/100
-	vpa := a.recommender.GetClusterState().Vpas[model.VpaID{Namespace: "", VpaName: ""}]
+	vpa := a.recommender.GetClusterState().Vpas[model.VpaID{Namespace: "", VpaName: "my-app-vpa"}] //TODO fix hardcoded vpa name in an issue https://github.com/pivotal/skenario/issues/114
```
Comment (josephburnett) [inline plugin-k8s-vpa/pkg/plugin/plugin.go:286]: Tip: You can just say `// TODO(#114): fix hardcoded vpa name`.

File: sim/pkg/model/autoscaler_ticktock.go [hunk 2]
```diff
@@ -62,6 +63,16 @@ func (asts *autoscalerTicktockStock) Add(entity simulator.Entity) error {
 	currentTime := asts.env.CurrentMovementTime()
 
 	asts.cluster.RecordToAutoscaler(&currentTime)
+
+	asts.adjustHorizontally(&currentTime)
+	asts.adjustVertically(&currentTime)
+
+	//calculate CPU utilization
```
Comment (josephburnett) [inline sim/pkg/model/autoscaler_ticktock.go:70]: I don't think you need this comment.  ;)

File: sim/pkg/model/autoscaler_ticktock.go [hunk 3]
```diff
@@ -98,13 +109,55 @@ func (asts *autoscalerTicktockStock) Add(entity simulator.Entity) error {
...
+func (asts *autoscalerTicktockStock) adjustVertically(currentTime *time.Time) {
+	recommendedPodResources, err := asts.env.Plugin().VerticalRecommendation(currentTime.UnixNano())
+	if err != nil {
+		panic(err)
+	}
+	//create a map podName -> []recommendedPodResources
+	podToRecommendations := make(map[string][]*proto.RecommendedPodResources, 0)
+	for _, recommendation := range recommendedPodResources {
+		podToRecommendations[recommendation.PodName] = append(podToRecommendations[recommendation.PodName], recommendation)
+	}
```
Comment (josephburnett) [inline sim/pkg/model/autoscaler_ticktock.go:123]: Does VPA provide recommendations for each pod?  I thought it was per container.  So if a Deployment template has pod with 2 containers, VPA would provide 2 recommendations.  No matter how many pods there are in the deployment.

@bskiba to confirm.

I see this code is consistent with the API: https://github.com/pivotal/skenario/blob/master/plugin/pkg/skplug/proto/skplug.proto#L61-L67 where I would expect `pod` to be `container`.  Sorry I didn't catch this earlier.

Comment (bskiba) [reply sim/pkg/model/autoscaler_ticktock.go:123]: Yes, VPA recommends per container. However, around the design phase there was an assumption that we simplify and assume always 1 container per pod (https://docs.google.com/document/d/1bD6q32JrqPetZniOe-VMYkm29Hh8-WhNEyCIOipD3TY/edit?disco=AAAAJ-e3U_I)

File: plugin-k8s-vpa/pkg/plugin/plugin.go (hunk 1 above)
Comment (juliababkina) [reply plugin-k8s-vpa/pkg/plugin/plugin.go:286]: Fixed it

File: sim/pkg/model/autoscaler_ticktock.go (hunk 2 above)
Comment (juliababkina) [reply sim/pkg/model/autoscaler_ticktock.go:70]: Removed it

File: sim/pkg/model/autoscaler_ticktock.go (hunk 3 above)
Comment (josephburnett) [reply sim/pkg/model/autoscaler_ticktock.go:122]: To be clear, if a Deployment has 100 pods, then VPA will produce 100 recommendations?

Comment (bskiba) [reply sim/pkg/model/autoscaler_ticktock.go:122]: No, just one per deployment.

File: sim/pkg/model/autoscaler_ticktock.go [hunk 4]
```diff
@@ -98,13 +109,55 @@ func (asts *autoscalerTicktockStock) Add(entity simulator.Entity) error {
...
 
-	return nil
+	//Iterate through replicas
+	pods := asts.cluster.ActiveStock().EntitiesInStock()
+	for _, pod := range pods {
+		//We have recommendations for the replica
+		for _, recommendation := range podToRecommendations[string((*pod).Name())] {
```
Comment (josephburnett) [inline sim/pkg/model/autoscaler_ticktock.go:128]: You should do away with this for loop.  Just apply the same recommendations to all pods.

File: plugin-k8s-vpa/go.mod [hunk 5]
```diff
@@ -41,6 +41,7 @@ replace (
 	k8s.io/sample-apiserver => k8s.io/sample-apiserver v0.17.9
 )
 
+// Checkout https://github.com/autoscaler/autoscaler.git branch plugin
```
Comment (josephburnett) [inline plugin-k8s-vpa/go.mod:44]: Should be `kubernetes/autoscaler`.

Comment (josephburnett) [reply plugin-k8s-vpa/go.mod:44]: But actually it should be `juliababkina/autoscaler`.

**Pull request merged**